CONTRIBUTING.md

```diff
@@ -0,0 +1,113 @@
+also check that you are using the latest version of Merlin and that a similar
+issue as not already been submitted.
+
+## Code contributions
+
+### Styleguides
+
+As of today, Merlin's codebase does not use a code formatter. When contributing
+code to an existing module, one should adopt the style of the surrounding code.
+Please keep lines under 80 characters.
+
+We plan to move the codebase to ocamlformat in a near future.
+
+Changes unrelated to the issue addressed by a PR should be made in a separate
+PR. Additionally, formatting changes in parts of the code not concerned by a
+specific PR should be proposed in another PR.
+
+### Menhir version
+
+Merlin promotes the generated Menhir parser in its sources. This is done to
+avoid depending on Menhir when installing Merlin. However this also means that
+unnecessary diff will appear when the parser gets re-generated by a different
+version of Menhir. To remove this diff please use version `20201216`:
+```bash
+$ opam pin menhir 20201216
+```
+The generated parser file should only be commited if there is an actual change
+in the grammar.
+
+### Repository organization
+
+
+#### `/src/ocaml`
+Vendored OCaml typer. Also include Merlin's incremental parser and other
+utilites.
+
+#### `/src/kernel`
+The backbone of merlin. Contains configuration utilities and pipeline
+management. The pipeline describe the compilation workflow: parsing, ppxing and
+typing.
+
+#### `/src/frontend/query_protocol.ml`
+The type description of Merlin's protocol as documented in
+[PROTOCOL.md](https://github.com/ocaml/merlin/blob/master/doc/dev/PROTOCOL.md)
+
+#### `/src/frontend/query_commands.ml`
+This file contains the dispatcher: a large pattern matching that answers
+Merlin's queries. Most of the time answering a query begins with getting the
+Typedtree from the pipeline and calling the correct analysis functions.
+
+#### `/src/analysis`
+Contains most of Merlin logic required to answer the queries.
+
+#### `/src/dot-protocol`
+The configuration protocol that configuration servers such as
+`dot-merlin-reader` or `dune ocaml-merlin` should implement.
+
+#### `/src/dot-merlin-reader`
+The historical documentation server of Merlin. Tt reads `.merlin` configuration
+files whose syntax is [documented in the
+wiki](https://github.com/ocaml/merlin/wiki/Project-configuration).
+
+#### `/src/frontend/ocamlmerlin`
+The standard Merlin frontend, composed of a long-running server and a client.
+This is the frontend used by the test-suite and standard emacs and vim plugins.
+For an alternative frontend, see
+[ocaml-lsp-server](https://github.com/ocaml/ocaml-lsp/)
+
+#### `/tests/test-dirs`
+Merlin's behavioral test-suite. We rely on Dune's cram test framework. Ideally,
+any opened issue should be accompanied by a test with a reproduction. When
+working on a fix for an issue, please have a first commit with the test showing
+the issue, before commits that fix the issue and update the test result
+accordingly.
```

Comment:
The comment that we appreciate a test-driven approach for PRs (everything from "Ideally," on) can be missed very easily here. What do you think about moving it to the _Styleguides_ section instead of having it here in _Repository Organization_?